### PR TITLE
[jit] Make `nn.Transformer` TorchScript compatible

### LINKED ISSUE #24173: Transformer model seems not supported in TorchScript?
## 🐛 Bug
```python
Python builtin <built-in function hasattr> is currently not supported in Torchscript:
at /home/chejin/anaconda3/envs/pytorch-gpu/lib/python3.6/site-packages/torch/nn/modules/activation.py:759:12
        - key_padding_mask: :math:`(N, S)`, ByteTensor, where N is the batch size, S is the source sequence length.
        - attn_mask: :math:`(L, S)` where L is the target sequence length, S is the source sequence length.

        - Outputs:
        - attn_output: :math:`(L, N, E)` where L is the target sequence length, N is the batch size,
          E is the embedding dimension.
        - attn_output_weights: :math:`(N, L, S)` where N is the batch size,
          L is the target sequence length, S is the source sequence length.
        """
        if hasattr(self, '_qkv_same_embed_dim') and self._qkv_same_embed_dim is False:
           ~~~~~~~ <--- HERE
            return F.multi_head_attention_forward(
                query, key, value, self.embed_dim, self.num_heads,
                self.in_proj_weight, self.in_proj_bias,
                self.bias_k, self.bias_v, self.add_zero_attn,
                self.dropout, self.out_proj.weight, self.out_proj.bias,
                training=self.training,
                key_padding_mask=key_padding_mask, need_weights=need_weights,
                attn_mask=attn_mask, use_separate_proj_weight=True,
                q_proj_weight=self.q_proj_weight, k_proj_weight=self.k_proj_weight,'__module__.__torch__.torch.nn.modules.activation.MultiheadAttention.forward' is being compiled since it was called from '__modul
e__.__torch__.torch.nn.modules.transformer.TransformerEncoderLayer.forward'
```
## Environment

Collecting environment information...
PyTorch version: 1.2.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.10.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti

Nvidia driver version: 430.26
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.16.2
[pip3] torchtext==0.3.1
[conda] _tflow_select             2.3.0                       mkl    defaults
[conda] blas                      1.0                         mkl    defaults
[conda] cuda100                   1.0                           0    pytorch
[conda] mkl                       2019.4                      243    defaults
[conda] mkl_fft                   1.0.12           py36ha843d7b_0    defaults
[conda] mkl_random                1.0.2            py36hd81dba3_0    defaults
[conda] pytorch                   1.2.0           py3.6_cuda10.0.130_cudnn7.6.2_0    pytorch
[conda] tensorflow                1.14.0          mkl_py36h2526735_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] tensorflow-base           1.14.0          mkl_py36h7ce6ba3_0    defaults

### PR DESCRIPTION
**Pull request by driazati**
This makes `nn.Transformer` usable from TorchScript. It preserves backwards compatibility via `__setstate__` on the encoder/decoder.

Fixes #24173

Differential Revision: [D18124753](https://our.internmc.facebook.com/intern/diff/18124753/)

### CODE REVIEW

**Comment by zhangguanheng66** [comment]
Add BC breaking tag for release note. @gchanan

**Comment by gchanan** [comment]
Remember to add label "topic: bc-breaking" to such changes as well :).

File: torch/nn/modules/transformer.py [hunk 1]
```diff
@@ -171,11 +173,10 @@ def forward(self, src, mask=None, src_key_padding_mask=None):
         """
         output = src
 
-        for i in range(self.num_layers):
```
Comment (zhangguanheng66) [inline torch/nn/modules/transformer.py]: So `range` is not jitable as well?

File: torch/nn/modules/transformer.py [hunk 2]
```diff
@@ -171,11 +173,10 @@ def forward(self, src, mask=None, src_key_padding_mask=None):
...
-            output = self.layers[i](output, src_mask=mask,
-                                    src_key_padding_mask=src_key_padding_mask)
+        for mod in self.layers:
+            output = mod(output, src_mask=mask, src_key_padding_mask=src_key_padding_mask)
 
-        if self.norm:
+        if self.norm is not None:
```
Comment (zhangguanheng66) [inline torch/nn/modules/transformer.py]: Is this change due to `jitable`?

File: torch/nn/modules/transformer.py [hunk 3]
```diff
@@ -196,16 +197,17 @@ class TransformerDecoder(Module):
         >>> tgt = torch.rand(20, 32, 512)
         >>> out = transformer_decoder(tgt, memory)
     """
+    __constants__ = ['norm']
 
     def __init__(self, decoder_layer, num_layers, norm=None):
         super(TransformerDecoder, self).__init__()
         self.layers = _get_clones(decoder_layer, num_layers)
-        self.num_layers = num_layers
```
Comment (zhangguanheng66) [inline torch/nn/modules/transformer.py]: This is a BC breaking

**Review by zhangguanheng66 — DISMISSED** [review]

File: torch/nn/modules/transformer.py (hunk 1 above)
Comment (driazati) [reply torch/nn/modules/transformer.py]: `ModuleList`s can't be indexed, but can be used with `for-in` loops. This should be the same

File: torch/nn/modules/transformer.py (hunk 2 above)
Comment (driazati) [reply torch/nn/modules/transformer.py]: Right the check has to be that the value is not None so this branch doesn't get compiled if there is no code to call (i.e. if `norm` is `None`)

File: torch/nn/modules/transformer.py (hunk 3 above)
Comment (driazati) [reply torch/nn/modules/transformer.py]: This seems like redundant info since someone can just do `len(self.layers)` to get the same value. Since this change is BC-breaking anyways I think this cleanup is warranted

**Review by zhangguanheng66 — COMMENTED** [review]
Do you need tests to cover the scriptable transformer?

**Review by zhangguanheng66 — COMMENTED** [review]
We are thinking about the `_load_from_state_dict` func to avoid BC breaking. See the PR I added. https://github.com/pytorch/pytorch/pull/29001

**Comment by driazati** [comment]
By the time 1.4 is released there will have been a whole minor version which I think is enough time to make a BC-breaking change here on loading old models. Either way it'd be good to settle on something, the first version of these PRs (this and #28555) had the proper BC-maintaining changes which I reverted based on our discussions.

**Comment by zhangguanheng66** [comment]
> By the time 1.4 is released there will have been a whole minor version for people to update their models, which I think is enough time to make a BC-breaking change here on loading old models.
> 
> Either way it'd be good to settle on something, the first version of these PRs (this and #28555) had the proper BC-maintaining changes which I reverted based on our discussions.

Yes. Based on our discussions, we agreed to remove `hasattr` func, which introduces BC breaking. Adding `_load_from_state_dict` could avoid such breaking while your current version still works.

Since I have the PR (https://github.com/pytorch/pytorch/pull/28555) to add `_load_from_state_dict` func for MultiheadAttention, I will add `_load_from_state_dict` func to nn.Transformer there, as well.

File: torch/nn/modules/transformer.py (hunk 3 above)
Comment (zhangguanheng66) [reply torch/nn/modules/transformer.py]: Since I'm targeting to cover BC, please add those two attributes back and I could approve it soon.

**Comment by zhangguanheng66** [comment]
Please consider to overwrite `__setstate__` func to avoid BC breaking. https://github.com/pytorch/pytorch/pull/29001.
Once it's done, we should remove the BC-breaking flag.

File: torch/nn/modules/transformer.py [hunk 4]
```diff
@@ -355,10 +366,7 @@ def forward(self, tgt, memory, tgt_mask=None, memory_mask=None,
                                    key_padding_mask=memory_key_padding_mask)[0]
         tgt = tgt + self.dropout2(tgt2)
         tgt = self.norm2(tgt)
-        if hasattr(self, "activation"):
```
Comment (driazati) [inline torch/nn/modules/transformer.py]: Since this gets patched up in `__setstate__`, the other case is not necessary

File: torch/nn/modules/transformer.py [hunk 5]
```diff
@@ -151,14 +152,15 @@ class TransformerEncoder(Module):
         >>> src = torch.rand(10, 32, 512)
         >>> out = transformer_encoder(src)
     """
+    __constants__ = ['norm']
 
     def __init__(self, encoder_layer, num_layers, norm=None):
         super(TransformerEncoder, self).__init__()
         self.layers = _get_clones(encoder_layer, num_layers)
-        self.num_layers = num_layers
```
Comment (zhangguanheng66) [inline torch/nn/modules/transformer.py]: nit: this will be a BC breaking. Prefer not to remove it since we won't claim a BC breaking for this module. Will this block a jitable transformer?

Comment (driazati) [reply torch/nn/modules/transformer.py]: I re-added it in the latest commit

File: torch/nn/modules/transformer.py [hunk 6]
```diff
@@ -332,8 +338,14 @@ def __init__(self, d_model, nhead, dim_feedforward=2048, dropout=0.1, activation
 
         self.activation = _get_activation_fn(activation)
 
+    def __setstate__(self, state):
+        if 'activation' not in state:
+            state['activation'] = _identity
```
Comment (zhangguanheng66) [inline torch/nn/modules/transformer.py]: The default one is `relu`.

**Review by zhangguanheng66 — CHANGES_REQUESTED** [review]
One more comment. Thanks.

File: torch/nn/modules/transformer.py (hunk 6 above)
Comment (driazati) [reply torch/nn/modules/transformer.py]: Fixed in the latest commit

File: torch/nn/modules/transformer.py (hunk 3 above)
Comment (zhangguanheng66) [reply torch/nn/modules/transformer.py]: `self.num_layers` is added to Encoder but not Decoder.

**Review by zhangguanheng66 — APPROVED** [review]
self.num_layers is added to Encoder but not Decoder. Please add it to Decoder and the PR is ready to merge. Thanks.
